[BUG]: Upgrade to sktime 0.12.0 breaks the unit tests

### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the develop branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@develop).


### Issue Description

Reported by @Yard1 , example here: https://github.com/pycaret/pycaret/runs/6997809658?check_suite_focus=true. Reverting back to 0.11.4 fixes the unit tests (https://github.com/pycaret/pycaret/pull/2672). This needs more evaluation.


### Reproducible Example

```python
Check unit tests
```


### Expected Behavior

Unit tests should pass

### Actual Results

```python-traceback
2022-06-22T06:15:26.7035044Z =========================== short test summary info ============================
2022-06-22T06:15:26.7035346Z FAILED pycaret/tests/test_time_series_base.py::test_create_predict_finalize_model[arima-fh4]
2022-06-22T06:15:26.7035543Z FAILED pycaret/tests/test_time_series_exogenous.py::test_create_tune_predict_finalize_model
2022-06-22T06:15:26.7035818Z FAILED pycaret/tests/test_time_series_indices.py::test_time_series_indices - ...
2022-06-22T06:15:26.7036084Z FAILED pycaret/tests/test_time_series_metrics.py::test_add_custom_metric - Ty...
2022-06-22T06:15:26.7036261Z FAILED pycaret/tests/test_time_series_preprocess.py::test_no_transform_noexo
2022-06-22T06:15:26.7036528Z FAILED pycaret/tests/test_time_series_preprocess.py::test_no_transform_exo - ...
2022-06-22T06:15:26.7036781Z ERROR pycaret/tests/test_time_series_blending.py::test_blend_model[voting] - ...
2022-06-22T06:15:26.7037050Z ERROR pycaret/tests/test_time_series_blending.py::test_blend_model[mean] - Ty...
2022-06-22T06:15:26.7037320Z ERROR pycaret/tests/test_time_series_blending.py::test_blend_model[median] - ...
2022-06-22T06:15:26.7037587Z ERROR pycaret/tests/test_time_series_blending.py::test_blend_model_predict - ...
2022-06-22T06:15:26.7037748Z = 6 failed, 435 passed, 1 skipped, 6373 warnings, 4 errors in 3277.61s (0:54:37) =
```


### Installed Versions

<details>
3.0.0rc on develop branch
</details>

--- a/pycaret/internal/pycaret_experiment/pycaret_experiment.py
+++ b/pycaret/internal/pycaret_experiment/pycaret_experiment.py
@@ -536,8 +536,13 @@ class _PyCaretExperiment:
             else:
                 return self.train_transformed
         else:
+            # Use fully trained pipeline to get the requested data
             return pd.concat(
-                [*_pipeline_transform(pipeline=self.pipeline, y=self.y, X=self.X)],
+                [
+                    *_pipeline_transform(
+                        pipeline=self.pipeline_fully_trained, y=self.y, X=self.X
+                    )
+                ],
                 axis=1,
             )
 
@@ -553,6 +558,7 @@ class _PyCaretExperiment:
             else:
                 return self.X_train_transformed
         else:
+            # Use pipeline trained on training data only to get the requested data
             # In time series, the order of arguments and returns may be reversed.
             return pd.concat(
                 [
@@ -572,15 +578,19 @@ class _PyCaretExperiment:
                 axis=1,
             )
         else:
+            # When transforming the test set, we can and should use all data before that
             # In time series, the order of arguments and returns may be reversed.
-            return pd.concat(
+            all_data = pd.concat(
                 [
                     *_pipeline_transform(
-                        pipeline=self.pipeline, y=self.y_test, X=self.X_test
+                        pipeline=self.pipeline_fully_trained,
+                        y=self.y,
+                        X=self.X,
                     )
                 ],
                 axis=1,
             )
+            return all_data.loc[self.y_test.index]
 
     @property
     def X_transformed(self):
@@ -591,8 +601,11 @@ class _PyCaretExperiment:
             else:
                 return self.X_train_transformed
         else:
+            # Use fully trained pipeline to get the requested data
             # In time series, the order of arguments and returns may be reversed.
-            return _pipeline_transform(pipeline=self.pipeline, y=self.y, X=self.X)[1]
+            return _pipeline_transform(
+                pipeline=self.pipeline_fully_trained, y=self.y, X=self.X
+            )[1]
 
     @property
     def y_transformed(self):
@@ -600,8 +613,11 @@ class _PyCaretExperiment:
         if self._ml_usecase != MLUsecase.TIME_SERIES:
             return pd.concat([self.y_train_transformed, self.y_test_transformed])
         else:
+            # Use fully trained pipeline to get the requested data
             # In time series, the order of arguments and returns may be reversed.
-            return _pipeline_transform(pipeline=self.pipeline, y=self.y, X=self.X)[0]
+            return _pipeline_transform(
+                pipeline=self.pipeline_fully_trained, y=self.y, X=self.X
+            )[0]
 
     @property
     def X_train_transformed(self):
@@ -616,6 +632,7 @@ class _PyCaretExperiment:
             else:
                 return self.pipeline.transform(self.X_train, filter_train_only=False)
         else:
+            # Use pipeline trained on training data only to get the requested data
             # In time series, the order of arguments and returns may be reversed.
             return _pipeline_transform(
                 pipeline=self.pipeline, y=self.y_train, X=self.X_train
@@ -629,7 +646,9 @@ class _PyCaretExperiment:
         else:
             # In time series, the order of arguments and returns may be reversed.
             # When transforming the test set, we can and should use all data before that
-            _, X = _pipeline_transform(pipeline=self.pipeline, y=self.y, X=self.X)
+            _, X = _pipeline_transform(
+                pipeline=self.pipeline_fully_trained, y=self.y, X=self.X
+            )
 
             if X is None:
                 return None
@@ -646,6 +665,7 @@ class _PyCaretExperiment:
                 filter_train_only=False,
             )[1]
         else:
+            # Use pipeline trained on training data only to get the requested data
             # In time series, the order of arguments and returns may be reversed.
             return _pipeline_transform(
                 pipeline=self.pipeline, y=self.y_train, X=self.X_train
@@ -659,5 +679,7 @@ class _PyCaretExperiment:
         else:
             # In time series, the order of arguments and returns may be reversed.
             # When transforming the test set, we can and should use all data before that
-            y, _ = _pipeline_transform(pipeline=self.pipeline, y=self.y, X=self.X)
+            y, _ = _pipeline_transform(
+                pipeline=self.pipeline_fully_trained, y=self.y, X=self.X
+            )
             return y.loc[self.y_test.index]

--- a/pycaret/tests/time_series_test_utils.py
+++ b/pycaret/tests/time_series_test_utils.py
@@ -316,7 +316,7 @@ def _return_data_big_small():
     """Returns one dataset with 144 data points and one with < 12 data points"""
     data = get_data("airline")
     data = data - 400
-    data_small = data[:12]  # 11 data points
+    data_small = data[:11]  # 11 data points
     datasets = [data, data_small]
 
     return datasets

--- a/pycaret/utils/time_series/__init__.py
+++ b/pycaret/utils/time_series/__init__.py
@@ -72,8 +72,12 @@ def _get_diffs(data: pd.Series, lags_list: List[Any]) -> List[pd.Series]:
     List[pd.Series]
         List of differences per the lags_list
     """
-
-    diffs = [Differencer(lags=lags).fit_transform(data) for lags in lags_list]
+    # Default na_handling changed in sktime 0.13.0 (compared to 0.11.x).
+    # https://www.sktime.org/en/latest/changelog.html#id3. Hence hard coding.
+    diffs = [
+        Differencer(lags=lags, na_handling="drop_na").fit_transform(data)
+        for lags in lags_list
+    ]
     return diffs
 
 

--- a/pycaret/utils/time_series/forecasting/pipeline.py
+++ b/pycaret/utils/time_series/forecasting/pipeline.py
@@ -159,6 +159,9 @@ def _get_imputed_data(
     Tuple[pd.Series, Optional[pd.DataFrame]]
         Imputed y and X values respectively
     """
+    # Since we are fitting (fit_transform) with new data, need to make sure
+    # we do not have side effects on pipeline
+    pipeline_ = deepcopy(pipeline)
 
     if X is None:
         # No exogenous variables
@@ -168,16 +171,16 @@ def _get_imputed_data(
         # Exogenous variables present
         X_imputed = X.copy()
 
-        if _transformations_present_X(pipeline):
-            transformer_pipeline_X = pipeline.steps_[0][1]
+        if _transformations_present_X(pipeline_):
+            transformer_pipeline_X = pipeline_.steps_[0][1]
             for _, transformer_X in transformer_pipeline_X.steps_:
                 if isinstance(transformer_X, Imputer):
                     X_imputed = transformer_X.fit_transform(X, y)
                     continue
 
     y_imputed = y.copy()
-    if _transformations_present_y(pipeline):
-        transformer_pipeline_y = pipeline.steps_[-1][1].steps_[0][1]
+    if _transformations_present_y(pipeline_):
+        transformer_pipeline_y = pipeline_.steps_[-1][1].steps_[0][1]
         for _, transformer_y in transformer_pipeline_y.steps_:
             if isinstance(transformer_y, Imputer):
                 y_imputed = transformer_y.fit_transform(y)

--- a/requirements.txt
+++ b/requirements.txt
@@ -28,6 +28,6 @@ plotly-resampler>=0.7.2.2
 
 # Time-series
 statsmodels>=0.12.1
-sktime~=0.11.4
+sktime~=0.13.0
 tbats>=1.1.0
 pmdarima>=1.8.0

